OEL-2766: Allow users to subscribe to events.

brummbar opened the pull request.

brummbar committed:
OEL-2766: Move entries to correct info file.

brummbar committed:
OEL-2766: Require oe_subscriptions.

brummbar committed:
OEL-2766: Create a message template for event updates.

brummbar committed:
OEL-2766: Delete unneeded flags upon install.

brummbar committed:
OEL-2766: Allow to subscribe to events.

brummbar committed:
OEL-2766: Render subscribe links.

brummbar committed:
OEL-2766: No need to export configs with default value.

brummbar committed:
OEL-2766: Fix dependencies.

brummbar committed:
OEL-2766: Move unused flag deletion to an install task.

brummbar committed:
OEL-2766: Sort config entries.

brummbar committed:
OEL-2766: Add symfony_mailer dependency.

brummbar committed:
OEL-2766: Export and use the email_html format.

brummbar committed:
OEL-2766: Add permissions to the anonymous subscriber role.

brummbar committed:
OEL-2766: Send a notification to subscribers when an event is updated.

brummbar committed:
OEL-2766: Add update path.

brummbar committed:
OEL-2766: Core module has been dropped.

brummbar committed:
OEL-2766: Export the configured flag.

brummbar committed:
OEL-2766: The subscribe_node flag is imported by flag module for existing installations.

brummbar committed:
OEL-2766: Enable symfony_mailer.

brummbar committed:
OEL-2766: Update code with changes from oe_whitelabel branches.

brummbar committed:
OEL-2766: Allow authenticated users to subscribe to nodes.

brummbar committed:
OEL-2766: Update dependencies.

brummbar committed:
OEL-2766: Reuse the whitelabel media trait as much as possible.

brummbar committed:
OEL-2766: Add a trait to create nodes.

brummbar committed:
OEL-2766: Write test coverage for digest emails.

brummbar committed:
OEL-2766: Update Selenium version.

AaronGilMartinez commented on modules/oe_showcase_subscriptions/oe_showcase_subscriptions.info.yml:
```diff
@@ -0,0 +1,23 @@
+name: OpenEuropa Showcase Subscriptions
+type: module
+description: Provides user subscriptions to content types.
+core_version_requirement: ^9.4 || ^10
```
We can drop 9 support

AaronGilMartinez commented on modules/oe_showcase_subscriptions/oe_showcase_subscriptions.module:
```diff
@@ -0,0 +1,31 @@
+ * OE Showcase Subscriptions module.
+ */
+
+declare(strict_types = 1);
```
```suggestion
declare(strict_types=1);
```

AaronGilMartinez commented on modules/oe_showcase_subscriptions/tests/src/ExistingSite/EventSubscriptionTest.php:
```diff
@@ -0,0 +1,334 @@
+   */
+  public function testEventNotifications(): void {
+    $event = $this->createEventNode();
+    $this->markEntityForCleanup($event);
```
Not needed, create function already marks it

brummbar committed:
OEL-2766: Fix lock rebase.

brummbar committed:
OEL-2766: Fix tests.

AaronGilMartinez commented on composer.json:
```diff
@@ -50,15 +53,17 @@
+        "openeuropa/oe_subscriptions": "1.x-dev",
         "openeuropa/oe_webtools": "^1.23",
-        "openeuropa/oe_whitelabel": "^1.3.0"
+        "openeuropa/oe_whitelabel": "0.2767.202303041517"
```
@todo to be changed

brummbar committed:
OEL-2766: Modify test service to collect mails instantly.

brummbar committed:
OEL-2766: Do not reset time after tests.

Time is used as reference to trigger executions (e.g. update mails
or cron tasks) after a certain time has passed. Moving it back
will make those conditions impossible to hit unless we move the
time enough forward again.

brummbar committed:
OEL-2766: Improve comments.

brummbar committed:
OEL-2766: Make sure no mails are sent when digest is on for all users.

brummbar committed:
OEL-2766: Fix mail wait assertion message.

brummbar committed:
OEL-2766: Disable update module to avoid mails during cron execution in tests.

brummbar committed:
OEL-2766: Asserting mails in this test is out of scope.

brummbar committed:
OEL-2766: Reset time after tests run.

This could cause issues if any time was stored in state
during execution, but we will handle this in a followup
ticket.

brummbar committed:
OEL-2766: Fix update path.

AaronGilMartinez commented on modules/oe_showcase_subscriptions/tests/src/ExistingSite/EventSubscriptionTest.php:
```diff
@@ -0,0 +1,346 @@
+    // Move the time forward one day.
+    \Drupal::time()->setTime(strtotime('+1 day +1 hour', \Drupal::time()->getCurrentTime()));
+    $this->cronRun();
+    $this->waitUntilMailsAreCollected(1);
```
Maybe comment why we have waits?

AaronGilMartinez commented on modules/oe_showcase_subscriptions/tests/src/ExistingSite/EventSubscriptionTest.php:
```diff
@@ -0,0 +1,346 @@
+    $assert_session->pageTextContains('Event Second event - update 1 has been updated.');
+    // Update also the first event.
+    $this->drupalGet($event->toUrl('edit-form'));
+    $assert_session->fieldExists('Title')->setValue('Event update 4');
```
Nitpick: Can we use same nomemclature for both events? 'Ordinal event - update n'

AaronGilMartinez commented on modules/oe_showcase_subscriptions/tests/src/ExistingSite/EventSubscriptionTest.php:
```diff
@@ -0,0 +1,346 @@
+  /**
+   * Retrieves the URIs of links existing in a mail.
+   *
+   * @param \Drupal\symfony_mailer\Email $mail
```
readMail() returns \Symfony\Component\Mime\Email
Should we change this?

brummbar replied:
It says it returns that but actually collected emails are of type \Drupal\symfony_mailer\Email.

brummbar committed:
OEL-2766: Fix coding standards and review remarks.
